nuxeo-core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/jdbc/dialect/Dialect.java

```diff
@@ -371,13 +371,14 @@ protected void setToPreparedStatementString(PreparedStatement ps, int index, Ser
 
     public void setToPreparedStatementTimestamp(PreparedStatement ps, int index, Serializable value, Column column)
             throws SQLException {
-        Calendar cal = (Calendar) value;
-        Timestamp ts = cal == null ? null : new Timestamp(cal.getTimeInMillis());
-        ps.setTimestamp(index, ts, cal); // cal passed for timezone
+        Timestamp ts = getTimestampFromCalendar((Calendar) value);
+        // NOTE we don't pass the Calendar as third parameter to setTimestamp(), because
+        // this causes h2, Oracle and SQL Server to write incorrect dates
```

Comment:
AFAIU database drivers interpret differently the timezone when calculating the timestamp because of bad interface specification.